src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -0,0 +1,1305 @@
+            (oldProject, documentIds, _) =>
+            {
+                var newProject = oldProject.RemoveAnalyzerConfigDocuments(documentIds);
+                return (newProject, new CompilationAndGeneratorDriverTranslationAction.ProjectCompilationOptionsAction(newProject, isAnalyzerConfigChange: true));
+            });
+    }
+
+    /// <summary>
+    /// Core helper that takes a set of <see cref="DocumentInfo" />s and does the application of the appropriate documents to each project.
+    /// </summary>
+    /// <param name="documentInfos">The set of documents to add.</param>
+    /// <param name="addDocumentsToProjectState">Returns the new <see cref="ProjectState"/> with the documents added, and the <see cref="SolutionCompilationState.CompilationAndGeneratorDriverTranslationAction"/> needed as well.</param>
+    /// <returns></returns>
+    private SolutionCompilationState AddDocumentsToMultipleProjects<T>(
+        ImmutableArray<DocumentInfo> documentInfos,
+        Func<DocumentInfo, ProjectState, T> createDocumentState,
+        Func<ProjectState, ImmutableArray<T>, (ProjectState newState, CompilationAndGeneratorDriverTranslationAction translationAction)> addDocumentsToProjectState)
+        where T : TextDocumentState
+    {
+        if (documentInfos.IsDefault)
+        {
+            throw new ArgumentNullException(nameof(documentInfos));
+        }
+
+        if (documentInfos.IsEmpty)
+        {
+            return this;
+        }
+
+        // The documents might be contributing to multiple different projects; split them by project and then we'll process
+        // project-at-a-time.
+        var documentInfosByProjectId = documentInfos.ToLookup(d => d.Id.ProjectId);
+
+        var newCompilationState = this;
+
+        foreach (var documentInfosInProject in documentInfosByProjectId)
+        {
+            this.Solution.CheckContainsProject(documentInfosInProject.Key);
+            var oldProjectState = this.Solution.GetProjectState(documentInfosInProject.Key)!;
+
+            var newDocumentStatesForProjectBuilder = ArrayBuilder<T>.GetInstance();
+
+            foreach (var documentInfo in documentInfosInProject)
+            {
+                newDocumentStatesForProjectBuilder.Add(createDocumentState(documentInfo, oldProjectState));
+            }
+
+            var newDocumentStatesForProject = newDocumentStatesForProjectBuilder.ToImmutableAndFree();
+
+            var (newProjectState, compilationTranslationAction) = addDocumentsToProjectState(oldProjectState, newDocumentStatesForProject);
+
+            (var newSolutionState, _, _) = newCompilationState.Solution.ForkProject(
+                oldProjectState,
+                newProjectState,
+                // intentionally accessing this.Solution here not newSolutionState
+                newFilePathToDocumentIdsMap: this.Solution.CreateFilePathToDocumentIdsMapWithAddedDocuments(newDocumentStatesForProject));
```

Comment:
forking the solution state first.  note: to preserve semantics with the original code CreateFilePathToXXX is called on this.Solution instead of the solution state we're building up.

Reply:
I agree you have maintained the behavior, but I strongly suspect that original behavior is a bug, and in VS we never notice since we'll only ever see all the documents applying to a single project. I'm quite surprised we didn't have a test for that, so if you don't want to investigate as a part of this PR please do file a tracking bug for this.